pkg/client/ssa_client.go

```diff
@@ -0,0 +1,218 @@
+		return false, err
+	}
+
+	if err := prepareForSSA(obj, c.Client.Scheme()); err != nil {
+		return false, err
+	}
+
+	if config.skipIf != nil && config.skipIf(obj) {
+		return false, nil
+	}
+
+	updated := true
+
+	// NOTE: once the SSA migration is not needed anymore, read the orig conditionally only when config.determineUpdate is true.
+
+	// this cannot fail - we would have failed previously already because the scheme wouldn't have known the object's kind
+	origRo, _ := c.Client.Scheme().New(obj.GetObjectKind().GroupVersionKind())
+	orig := origRo.(client.Object)
+
+	if err := c.Client.Get(ctx, client.ObjectKeyFromObject(obj), orig); err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+
+	// NOTE: remove this once this code is used for all "apply-style" functionality and has updated all the objects in the cluster.
+	if isSsaMigrationNeeded(orig, c.NonSSAFieldOwner) {
+		if err := c.MigrateToSSA(ctx, orig); err != nil {
+			return false, err
+		}
+	}
+
+	if err := c.Client.Patch(ctx, obj, client.Apply, client.FieldOwner(c.FieldOwner), client.ForceOwnership); err != nil {
+		return false, fmt.Errorf("unable to patch '%s' called '%s' in namespace '%s': %w", obj.GetObjectKind().GroupVersionKind(), obj.GetName(), obj.GetNamespace(), err)
+	}
+
+	if config.determineUpdate {
+		updated = obj.GetGeneration() != orig.GetGeneration()
+	}
+
+	return updated, nil
+}
+
+func prepareForSSA(obj client.Object, scheme *runtime.Scheme) error {
+	gvk, err := apiutil.GVKForObject(obj, scheme)
+	if err != nil {
+		return err
+	}
+	obj.GetObjectKind().SetGroupVersionKind(gvk)
+	obj.SetManagedFields(nil)
+
+	return nil
+}
+
+// Apply applies the objects, ie, creates or updates them on the cluster
+// returns `true, nil` if at least one of the objects was created or modified,
+// `false, nil` if nothing changed, and `false, err` if an error occurred
+//
+// NOTE: this is only used in tests
+func (c *SsaApplyClient) Apply(ctx context.Context, toolchainObjects []client.Object, opts ...SsaApplyObjectOption) (bool, error) {
+	createdOrUpdated := false
+	for _, toolchainObject := range toolchainObjects {
+		result, err := c.ApplyObject(ctx, toolchainObject, opts...)
+		if err != nil {
+			return createdOrUpdated, errors.Wrapf(err, "unable to create resource of kind: %s, version: %s", toolchainObject.GetObjectKind().GroupVersionKind().Kind, toolchainObject.GetObjectKind().GroupVersionKind().Version)
+		}
+		createdOrUpdated = createdOrUpdated || result
+	}
+	return createdOrUpdated, nil
+}
+
+func (c *SsaApplyClient) MigrateToSSA(ctx context.Context, obj client.Object) error {
+	if err := csaupgrade.UpgradeManagedFields(obj, sets.New(c.NonSSAFieldOwner), c.FieldOwner); err != nil {
+		return err
+	}
+
+	return c.Client.Update(ctx, obj)
+}
+
+func isSsaMigrationNeeded(obj client.Object, nonSsaOwner string) bool {
+	for _, mf := range obj.GetManagedFields() {
+		if mf.Manager == nonSsaOwner {
```

Comment:
I'm wondering how this would be different compared to a case when a user deletes the managedFields completely. If it's the same, then we would need to keep the migration forever.

Reply:
never mind, I misread the code. it actually checks if the nonSsaOwner was used before or not.